docs/community/rust-coding-guidelines/rust-secure-coding-guidelines.md

```diff
@@ -0,0 +1,253 @@
+
+## Language Generalities
+
+### Unsafe Code
+
+Never use `unsafe` blocks unless as a last resort. Justify their use in a comment, detailing how the code is effectively safe to deploy [[unsafe code]](https://anssi-fr.github.io/rust-guide/04_language.html#unsafe-code).
+
+### Integer Overflows
+
+Refer to [coding-guidelines](./rust-coding-guidelines.md#integer-arithmetic).
+
+### Error Handling
+
+Use `Result<T, E>` and `Option<T>` for error handling instead of _unwrapping_ or _expecting_, to avoid panics [[error handling]](https://anssi-fr.github.io/rust-guide/04_language.html#error-handling) [[panic]](https://anssi-fr.github.io/rust-guide/04_language.html#panics) [[coding-style]](rust-coding-guidelines.md/#error-handling).
+
+### Assertions
+
+Prefer using `Result` and context-rich error handling over Rust's `assert!`, `assert_eq!`, and `assert_ne!` macros for enforcing invariants, reserving assertions for development and unrecoverable error scenarios.
+
+## Types Systems and Data Structures
+
+### Drop Trait
+
+Implement the `Drop` trait selectively and ensure it doesn't cause panics [[drop trait the destructor]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#drop-trait-the-destructor).
+
+- Implement the Drop trait selectively, only when necessary for specific destructor logic. It's mainly used for managing external resources or memory in structures like Box or Rc, often involving unsafe code and security-critical operations.
+- In a Rust secure development, the implementation of the `std::ops::Drop` trait
+  must not panic.
+- Do not rely on `Drop` trait in security material treatment after the use, use [zeroize](https://docs.rs/zeroize/latest/zeroize/#) to explicit destroy security material, e.g. private keys.
+
+### Send and Sync Traits
+
+Be cautious with manual implementations of `Send` and `Sync` traits [[Rustbook: typesystem]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#send-and-sync-traits) [[Rustbook: send and sync]](https://doc.rust-lang.org/nomicon/send-and-sync.html).
+Both traits are _unsafe traits_, i.e., the Rust compiler does not verify in any way that they are implemented correctly. The danger is real: an incorrect implementation may lead to **undefined behavior**.
+
+In the majority of scenarios, manual implementation is unnecessary. In Rust, nearly all primitive types intrinsically implement Send and Sync traits, and for a significant proportion of compound types, the Rust compiler automatically derives these implementations.
+
+### Comparison Traits
+
+Ensure the implementation of standard comparison traits respects documented invariants [[comparison traits partialeq eq partialord ord]](https://anssi-fr.github.io/rust-guide/06_typesystem.html#comparison-traits-partialeq-eq-partialord-ord).
```

Comment:
same question about relevancy - also what are the "documented invariants" and what does it mean within the aptos core context?

Reply:
Updated with a definition of invariants in Comparison Traits